Add implementation of Poseidon2

Al-Kindi-0 opened the pull request:
## Describe your changes
Closes #429 

One interesting thing regarding the benchmarks is that, on my laptop, I am seeing an up-to-2x slowdown between the cached and random benchmarks for Poseidon2. Would be good to know if this is architecture specific or not.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

plafer commented on miden-crypto/benches/hash.rs:
```diff
@@ -100,6 +100,52 @@ fn rpx256_sequential(c: &mut Criterion) {
+fn poseidon2_2to1(c: &mut Criterion) {
+    let v: [Word; 2] = [Poseidon2::hash(&[1_u8]), Poseidon2::hash(&[2_u8])];
+    c.bench_function("Poseidon2 2-to-1 hashing (cached)", |bench| {
+        bench.iter(|| Poseidon2::merge(black_box(&v)))
```
How does this implementation compare with HorizenLabs'?

Al-Kindi-0 replied:
I am getting `7.7344 µs` for their version vs `1.7 µs` in this PR for 1 call to the permutation. Probably it is due to their use of Arkworks generic finite field arithmetic, but not sure.

Al-Kindi-0 commented on miden-crypto/src/hash/poseidon2/mod.rs:
```diff
@@ -0,0 +1,659 @@
+    /// Applies Poseidon2 permutation to the provided state.
+    #[inline(always)]
+    pub fn apply_permutation(state: &mut [Felt; STATE_WIDTH]) {
+        // 1. Apply (external) linear layer to the input
+        Self::apply_matmul_external(state);
+
+        // 2. Apply initial external rounds to the state
+        Self::initial_external_rounds(state);
+
+        // 3. Apply internal rounds to the state
+        Self::internal_rounds(state);
+
+        // 4. Apply terminal external rounds to the state
+        Self::terminal_external_rounds(state);
```
Yes, that is indeed the case. The round constants are different even between the external rounds and this is indeed an additional complexity with respect to the arithmetization.

bobbinth commented on miden-crypto/benches/README.md:
```diff
@@ -5,39 +5,41 @@ In the Miden VM, we make use of different hash functions. Some of these are "tra
 * **Rescue Prime (RP)** as specified [here](https://eprint.iacr.org/2020/1143) and implemented [here](https://github.com/novifinancial/winterfell/blob/46dce1adf0/crypto/src/hash/rescue/rp64_256/mod.rs).
 * **Rescue Prime Optimized (RPO)** as specified [here](https://eprint.iacr.org/2022/1577) and implemented in this crate.
 * **Rescue Prime Extended (RPX)** a variant of the [xHash](https://eprint.iacr.org/2023/1045) hash function as implemented in this crate.
+* **Poseidon2** as specified [here](https://eprint.iacr.org/2023/1045) and implemented in this crate.
```
The paper link is off here as well.

adr1anh reviewed (approved):
Looks good to me!

bobbinth commented on miden-crypto/src/hash/algebraic_sponge/rescue/rpo/mod.rs:
```diff
@@ -0,0 +1,102 @@
+/// ## Hashing of empty input
+/// The current implementation hashes empty input to the zero digest [0, 0, 0, 0]. This has
+/// the benefit of requiring no calls to the RPO permutation when hashing empty input.
+pub type Rpo256 = AlgebraicSponge<Rpo256Permutation>;
+
+/// The Rescue Prime Optimized permutation function with 256-bit state size.
+pub struct Rpo256Permutation();
```
I'm a bit concerned about exporting `Rpo256` as a type alias of `AlgebraicSponge` - I think it may be leaking a bit too much about the internal details of how it is implemented. Maybe it is better to have `AlgebraicSponge` be fully transparent for the end user: it would be an internal trait, similar to how the `SparseMerkleTree` trait works.

Basically, we'd have the `AlgebraicSponge` be a `pub(crate)` trait. That looks something like:

```rust
pub(crate) trait AlgebraicSponge {
    fn apply_permutation(state: &mut [Felt; STATE_WIDTH]);

    fn hash(bytes: &[u8]) -> Word {
        todo!("implement")
    }

    fn merge(values: &[Word; 2]) -> Word {
        todo!("implement")
    }

    fn merge_many(values: &[Word]) -> Word {
        todo!("implement")
    }

    fn merge_with_int(seed: Word, value: u64) -> Word {
        todo!("implement")
    }

    fn hash_elements<E: FieldElement<BaseField = Felt>>(elements: &[E]) -> Word {
        todo!("implement")
    }
}
```

Then, we'd do something like:
```rust
pub struct Rpo256();

impl AlgebraicSponge for Rpo256 {
    fn apply_permutation(state: &mut [Felt; STATE_WIDTH]) {
        todo!("implement")
    }
}

impl Hasher for Rpo256 {
    const COLLISION_RESISTANCE: u32 = 128;
    
    type Digest = Word;

    fn hash(bytes: &[u8]) -> Self::Digest {
        <Self as AlgebraicSponge>::hash(bytes)
    }

    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
        <Self as AlgebraicSponge>::merge(values)
    }

    fn merge_many(values: &[Self::Digest]) -> Self::Digest {
        <Self as AlgebraicSponge>::merge_many(values)
    }

    fn merge_with_int(seed: Self::Digest, value: u64) -> Self::Digest {
        <Self as AlgebraicSponge>::merge_with_int(seed, value)
    }
}

impl ElementHasher for Rpo256 {
    type BaseField = Felt;

    fn hash_elements<E: FieldElement<BaseField = Self::BaseField>>(elements: &[E]) -> Self::Digest {
        <Self as AlgebraicSponge>::hash_elements(elements)
    }
}
```

It is a bit more boilerplate - but externally, our `Rpo256` looks exactly like it did before (and this would actually be a non-breaking change).

Or do you think this would cause complications somehow?

Al-Kindi-0 commented:
Note that Rust was complaining about two things:

1. `merge_in_domain` being not used. I went ahead and added a `#[allow(dead_code)]`
2. the following `/// [merge_with_int()](Rpo256::merge_with_int) are internally consistent. That is, computing` is a public documentation to a private item, which is not the case as far as I can see. I went ahead and added a `#[allow(rustdoc::private_intra_doc_links)]`

Al-Kindi-0 commented on miden-crypto/src/hash/algebraic_sponge/mod.rs:
```diff
@@ -0,0 +1,201 @@
+use core::ops::Range;
```
Added, let me know if we should add/elaborate more

Al-Kindi-0 commented on miden-crypto/src/hash/algebraic_sponge/mod.rs:
```diff
@@ -0,0 +1,201 @@
+// ALGEBRAIC SPONGE
+// ================================================================================================
+
+pub trait AlgebraicSponge {
```
done

plafer commented on miden-crypto/src/hash/algebraic_sponge/mod.rs:
```diff
@@ -0,0 +1,244 @@
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a hash of two digests and a domain identifier.
+    #[allow(dead_code)]
```
This this is a `pub(crate)` trait, do we ever expect to use this internally? Otherwise we're just maintaining dead code

bobbinth replied:
I'm not sure we need this to marked as `dead_code`. I think `Rpo256::merge_in_domain()` should just be public and then all dead code warnings go away. I'll make a commit to fix this.

Al-Kindi-0 replied:
Yep, missed that during the copy-paste. Thanks for fixing it!
